### DISCUSSION TITLE
IO defining/extending a device target?

### DISCUSSION THREAD

**Discussion by Templarian** [Moddable-OpenSource/moddable#927] [Q&A] [answered]
Wrote a [`sensor/Joycon`](https://github.com/Templarian/moddable/blob/joycon/modules/drivers/sensors/joycon/joycon.js) to handle the Joycon analog joystick.

Question: While writing this I was making an example, but couldn't find any guidance of how to define the device. Is this meant to extend the `import device from "embedded:provider/builtin";` or create a new device object entirely?

https://github.com/Templarian/moddable/blob/joycon/examples/drivers/sensors/joycon/main.js#L51

Off topic:
- Is this kind of "sensor" useful to make a PR for or is it too specific (could be renamed to Joystick)?
  - I ask since this ended up being throwaway work as I like the d-pad more for this project.
- I saw a comment `//@@ Move Button class to common module` is this replacing the old `Button` soon?

**Comment by phoddie** [comment, marked as answer]
Apologies for being a little slow to get to this.

The driver looks great! (For completeness on close it should close the two analog instances to free up the memory and release the pins.) 

> While writing this I was making an example, but couldn't find any guidance of how to define the device. Is this meant to extend the import device from "embedded:provider/builtin"; or create a new device object entirely?

What you've done here is reasonable. Building the device is often a bit ad-hoc when adding peripherals to an existing device. We did something [similar](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/io/tcp/mqtt/config.js) for our recent network protocol work (HTTP, WebSocket, etc). A host provider for an actual device with a built-in joystick would do this directly.

For sensors, the way we like to set-up the device is a bit different. More or less, the goal is to allow instantiation with something like:

```js
js = new device.sensor.Joycon;
```

A good example of that style is the [temperature and humidity sensor](https://github.com/Moddable-OpenSource/moddable/blob/b620dca3612efe5ae46c1e5cf9ef5ac26b363be7/build/devices/esp32/targets/m5paper/host/provider.js#L137-L148) built into the M5Paper. In your case, that might look something like this:

```js
device.sensor ?? = {};
device.sensor.Joycon = class {
	constructor(options) {
		new Joycon({
			...options,
			x: {
				io: device.io.Analog,
				pin: device.pin.joystickX
			},
			y: {
				io: device.io.Analog,
				pin: device.pin.joystickY
			}
		});
	}
};
```

You might not even define `joystickX` & `joystickY` on `device.pin` and just use constants here, since users of the class won't need them. Similarly, you don't need the additional properties on `device.Analog`.

> I saw a comment //@@ Move Button class to common module is this replacing the old Button soon?

That's the plan. We're moving to phase out pins (which is how the old Button is implemented) is favor of the ECMA-419 IO.

> Is this kind of "sensor" useful to make a PR for or is it too specific (could be renamed to Joystick)?

Yes! It would be great to have a PR for this so anyone who needs a basic analog joystick controller doesn't need to start over.
